00_basics/sc_dimensions.ipynb

```diff
@@ -0,0 +1,771 @@
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "-> 3"
+     ]
+    }
+   ],
+   "source": [
+    "a.size;"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "hawaiian-poverty",
+   "metadata": {},
+   "source": [
+    "Note also that due the circumstance that a set has no order we can also not access the *i*-th element of a set."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 20,
+   "id": "generic-consensus",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "ERROR: Message 'at' not understood.\n",
+      "RECEIVER:\n",
+      "Instance of Set {    (0x7f98d8263488, gc=C0, fmt=00, flg=00, set=02)\n",
+      "  instance variables [2]\n",
+      "    array : instance of Array (0x7f99101a8108, size=8, set=3)\n",
+      "    size : Integer 3\n",
+      "}\n",
+      "ARGS:\n",
+      "   Integer 0\n",
+      "\n",
+      "PROTECTED CALL STACK:\n",
+      "\tMeta_MethodError:new\t0x7f99183ac080\n",
+      "\t\targ this = DoesNotUnderstandError\n",
+      "\t\targ what = nil\n",
+      "\t\targ receiver = Set[ 3, 2, 1 ]\n",
+      "\tMeta_DoesNotUnderstandError:new\t0x7f99183ae040\n",
+      "\t\targ this = DoesNotUnderstandError\n",
+      "\t\targ receiver = Set[ 3, 2, 1 ]\n",
+      "\t\targ selector = at\n",
+      "\t\targ args = [ 0 ]\n",
+      "\tObject:doesNotUnderstand\t0x7f99181fb200\n",
+      "\t\targ this = Set[ 3, 2, 1 ]\n",
+      "\t\targ selector = at\n",
+      "\t\targ args = nil\n",
+      "\ta FunctionDef\t0x7f98d80f6a08\n",
+      "\t\tsourceCode = \"{ var result; \\\"**** JUPYTER ****\\\".postln; result = {a[0]}.value(); postf(\\\"-> %\n",
+      "\\\", result); \\\"\\\".postln;}\"\n"
+     ]
+    }
+   ],
+   "source": [
+    "a[0]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "blocked-weight",
+   "metadata": {},
+   "source": [
+    "## Working with matrices\n",
+    "\n",
+    "Lets start by implementing a function which will multiply two matrices which also checks if this is a valid operation."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 21,
+   "id": "gothic-thong",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "-> a Function"
+     ]
+    }
+   ],
+   "source": [
+    "~matrixMul = {|a, b|\n",
+    "    if(a[0].size == b.size, {\n",
+    "        a.collect({|d, i|\n",
+    "            (b.collect({|z| z}) * d).sum;\n",
+    "        });\n",
+    "    }, {\n",
+    "        \"error\".postln;\n",
+    "    });\n",
+    "};"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 22,
+   "id": "mobile-cycle",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "-> [ [ 4 ], [ 6 ] ]"
+     ]
+    }
+   ],
+   "source": [
+    "~matrixMul.(\n",
+    "    [[1, 1, 1], [1,1,3]],\n",
```

Comment:
this second matrix looks complicated. We didn't need any of that in the seminar. Maybe there is confustion about columns and rows?

Reply:
The original code was not working properly in my understanding

```supercollider
m = { |a, b|
    b.collect { |coefs| sum(a * coefs) }
};
m.(
    [[1, 1, 1], [1,1,3]],
    [[2],[1],[1]]
)
```
returns
```
-> [ [ 4, 4, 8 ], [ 2, 2, 4 ], [ 2, 2, 4 ] ]
```

But multiplying a 2x3 with a 3x1 matrix should result in a 2x1 matrix and not in a 3x3 matrix

Reply:
My nice bookshelf metaphor was completely wrong. What a shame, will have to explain it all differently again! This is why I didn't study math.

So matrix indices are *rows* first, columns second.

```
// these are then two *rows*
[
[1, 1, 1], 
[1, 1, 3]
]
```

But well, we speak of a stack of books then.

The formula above is then

```supercollider
 b.flop.collect { |coefs| sum(a * coefs) }
```

Reply:
It all doesn't make sense to me any more to do this as row vectors. The rows seem to have no real meaning for matrix multiplication?